rocky/katalogus/client.py

```diff
@@ -121,181 +144,200 @@ def __init__(self):
-        self.organization_uri = f"/v1/organisations/{organization}"
+        if error.response.status_code == codes.BAD_REQUEST and "Duplicate plugin" in error.response.text:
+            error_message = json.loads(error.response.text).get("detail")
+            raise DuplicatePluginError(error_message) from error
 
-    def organization_exists(self) -> bool:
-        response = self.session.get(self.organization_uri)
+        if error.response.status_code in [codes.FORBIDDEN, codes.NOT_FOUND]:
+            raise KATalogusNotAllowedError
 
-        return response.status_code != 404
+        raise KATalogusHTTPStatusError(error) from error
 
-    def create_organization(self, name: str):
-        response = self.session.post("/v1/organisations/", json={"id": self.organization, "name": name})
-        response.raise_for_status()
 
-        logger.info("Created organization", name=name)
+class KATalogusClient:
+    def __init__(self, base_uri: str):
+        self.session = httpx.Client(base_url=base_uri, event_hooks={"response": [verify_response]})
 
-    def delete_organization(self):
-        response = self.session.delete(self.organization_uri)
-        response.raise_for_status()
+    def health(self) -> ServiceHealth:
+        response = self.session.get("/health")
 
-        logger.info("Deleted organization", organization_code=self.organization)
+        return ServiceHealth.model_validate_json(response.content)
 
-    def get_plugins(self, **params) -> list[Plugin]:
+    def organization_exists(self, organization_code: str) -> bool:
         try:
-            response = self.session.get(f"{self.organization_uri}/plugins", params=params)
-            response.raise_for_status()
-        except httpx.HTTPStatusError as error:
-            raise KATalogusHTTPStatusError(error)
-        return [parse_plugin(plugin) for plugin in response.json()]
+            self.session.get(f"/v1/organisations/{organization_code}")
+        except KATalogusNotAllowedError:
+            return False
 
-    def get_plugin(self, plugin_id: str) -> Plugin:
-        plugin_id = quote(plugin_id)
-        response = self.session.get(f"{self.organization_uri}/plugins/{plugin_id}")
-        response.raise_for_status()
+        return True
 
-        return parse_plugin(response.json())
+    def create_organization(self, organization):
+        self.session.post("/v1/organisations/", json={"id": organization.code, "name": organization.name})
 
-    def get_plugin_schema(self, plugin_id: str) -> dict | None:
-        plugin_id = quote(plugin_id)
-        response = self.session.get(f"{self.organization_uri}/plugins/{plugin_id}/schema.json")
-        response.raise_for_status()
+        logger.info("Created organization", name=organization.name)
 
-        schema = response.json()
-        if not schema:
-            return None
+    def delete_organization(self, organization_code: str):
```

Comment:
The following methods should be available only on the Katalogus Class I'd say. We could then allow the user to instantiate a Katalogus for a given organization *only* if they have the correct rights / memberships, and make sure we have only one codepath to get there.

I'd expect a factory on the user object that allows for the instantiation of the Katalogus instance bound to the organization requested.

Reply:
I was also looking into passing a user to the base client and do permission checking. However the complexity involved is that the permissions are mostly defined per _method_, not per instance.  Do we check for `"tools.can_scan_organization"`? Or `"tools.can_set_katalogus_settings"`? Perhaps `"tools.can_enable_disable_boefje"` as well? I could add those checks on the method calls instead, then probably add a helper factory-method to the OrganizationView.

Currently, most permissions are defined at the View level, which we should maintain properly to treat these new checks as a fallback. (On a side note: having the option to check permissions before moving forward is in my eyes another benefit of not extending the KATalogusClient as a parent and give it a bit more responsibility, having different behavior than a client.)